src/foam/nanos/controller/WindowHash.js

```diff
@@ -23,17 +23,31 @@ foam.CLASS({
     function initArgs(args, ctx) {
       this.SUPER(args, ctx);
 
-      this.value$.sub(this.onValueChange);
       this.onPopState();
       this.window.onpopstate = this.onPopState;
+    },
+    function valueChanged(value, replaceHistoryState) {
+      if ( value )
+        this.value = value;
+
+      if ( replaceHistoryState )
+        this.window.history.replaceState(null, '', this.window.origin + '/#' + this.value);
+      else
+        this.updateHash();
+    },
+    function updateHash() {
+      if ( this.feedback_ ) return;
+
+        this.feedback_ = true;
+        this.window.location.hash = this.value;
+        this.feedback_ = false;
     }
   ],
 
   listeners: [
     function onPopState() {
       if ( this.feedback_ ) return;
```

Comment:
for some reason in `treviso`:
`this.window.onpopstate = this.onPopState` popstate event isn't fired. not sure if it should.

so that's why I added https://github.com/nanoPayinc/NANOPAY/pull/12822/files